Feature/directed attacks (Need Design Feedback)

MaxHT0x opened the pull request:
## Description:

closes #112
updated 10/25:

Players aim attacks by clicking a target location during attack initiation.

### How It Works

The system captures the clicked tile when a player initiates an attack and guides expansion through a priority system that balances four factors:

1. **Tile defensibility** - terrain difficulty and adjacent owned tiles
2. **Wave front progression** - earlier-discovered tiles take priority (linear growth)
3. **Directional bias** - tiles aligned with the click direction score higher (dot product alignment)
4. **Proximity bonus** - tiles near the click receive bonus priority (BFS distance decay)

Direction and proximity decay exponentially over time, creating natural "aim fade" as attacks progress.

### Configuration

Tune all parameters in `DefaultConfig.ts`:
- `attackDirectionWeight()`: 2.5 (direction influence)
- `attackMagnitudeWeight()`: 0.75 (proximity bonus)
- `attackTimeDecay()`: 20.0 ticks (fade rate)
- `attackDistanceDecayConstant()`: 25.0 tiles (proximity decay rate)
- `attackBFSMaxRadius()`: 200.0 tiles (prevents performance issues)
- `attackBFSDownsampleFactor()`: 10 (grid sampling interval)

### Performance

BFS optimization downsamples by 10x with a 200-tile radius limit:

### Implementation Details

- **Per-tile vectors:** Each border tile calculates its own direction to the click point (creates triangular convergence)
- **Connected-region BFS:** Traverses only target-owned land tiles (respects topology)
- **Exponential decay:** Direction and proximity fade naturally over attack duration
- **Backward compatible:** Attacks without clickTile work exactly as before






## Please complete the following:

- [x] I have added screenshots for all UI updates
- [x] I process any text displayed to the user through translateText() and I've added it to the en.json file
- [x] I have added relevant tests to the test directory
- [x] I confirm I have thoroughly tested these changes and take full responsibility for any bugs introduced

## Please put your Discord username so you can be contacted if a bug or regression is found:
stackk.

evanpelle commented:
One thing I'm bit hesitant when using distance is that clicking farther away from your border could increase the directional bias vs clicking right next to your border. I think ideally it should be more about the direction you click, and distance from border should matter.

nottirb commented:
This implementation is nearly identical to the first directed attack system I prototyped. I ran into several edge cases during testing and ended up scrapping it.

Here's what I observed:
 - Topological mismatch: Using euclidean magnitude is simple, but it doesn't respect terrain connectivity. For example, two borders can be geometrically close but separated by water, a river, or another player, causing attraction to topologically unreachable tiles.
 - Multiple border bias: Suppose you share multiple borders with an opponent, distance weighting tends to open deep penetration from all sides simultaneously. Since the attacks are thin yet penetrate deeply, it's pretty easy to collapse it and trigger a full annexation on the territory you attacked (making your directional attack functionally worthless).
 - Late game instability (post-MIRV): Because the terrain is heavily nuked and fragmented, weighting by raw distance causes the attack to tend toward the direction you clicked and dig itself into holes or isolated craters in the map. This essentially makes your expansion into heavily nuked land behave like DFS traversal (spider-like, starting/stopping randomly in disconnected patches of land).
 - Directional paradoxes: Clicking on one side of an enemy can trigger attacks from a completely different side, simply because that regions border happens to be geometrically close, despite being topologically further. (i.e. Russia want to attack eastern China, yet the attack starts from Western Russia because it happens to share a small 1-3 tile border which is geometrically closer)

After running into these issues, I scrapped the magnitude -based approach and was experimenting with the following:
 - Vector-aligned attacks: By comparing a local move vector (tile -> neighbor) with the intended attack vector(tile -> click), we can make the entire front as a whole lean toward the clicked direction instead of tunneling along a single vulnerable path. There is some visible square-grid quantization problems, but because we take the dot product of a discrete move vector with a continuous attack vector it interpolates rather smoothly and resolved the directional paradoxes and the multiple border bias problem. 
 - Pathfinding based propagation (I tried variants of BFS, DFS, A*, and the current system boat attacks use) to preserve topological awareness. These attacks obviously handled the topological mismatch and late game instability problems beautifully, but the result was overly robotic and end up being pretty expensive when used aggressively across entire borders.

Happy to pass over this PR to you since I'm working on it slowly. If we want to preserve OpenFront's fluidity, a hybrid model should work well I think and was my current path of exploration:
 - Using what you call the defensibility weight, since it biases the expansion in a way the players are familiar to
 - When an attack begins, compute full BFS distances for each tile, but only use them as a soft preference. This preserves topological correctness while avoiding robotic straight-line routing. It also means your attack will be consistently biased around the same area, even if it gets cut off. This also means your enemies can strategically stop your directional attack by making the (initially) topologically best path super expensive, forcing you to attack a new spot or retreat.
 - Continuously vector-align the attacks so that closest & largest fronts as a whole move towards the target instead of tunneling on a few tile-long paths.
 - Add distance decay to prevent long-range domination when clicking in the middle of a target
 - Add time decay so that while the attack is initially concentrated towards the target, it slowly fades back into the current attack method we know and love

Doing this, while quite technically involved, lets the attacks behave much like they do now, but with a clear, temporary directional bias which fades over time and can be strategically countered. It preserves OpenFront's familiar attack fluidity while adding a layer strategic depth (and doesn't require an additional slider). 

The attacks start of guided and cohesive, then organically revert to standard expansion. This system is topologically aware, resolves all 4 of the above issues, the concern @evanpelle raised, and is computationally inexpensive: you only need to do a single BFS pass into a hashmap, and the eight discrete move vectors can be precomputed for all tiles. This ends up not only being cheaper than the current boat implementation, yet far more strategically interesting. It also has a nice side effect in that clicking near the center naturally drives multiple coordinated fronts forward cohesively due to the underling math.

MaxHT0x commented:
> nottrib snip


Thank you for the very insightful comment, also very sorry i didn't notice! You seem much more passionate on this issue then me to be honest, so if you want to complete it since you already started please let me know. 

I believe with my last commit i am at the vector alignment approach you mentioned. It does seem much better than all previous attempts I made. 

current implmentation:
- Using dot product of (attacker > click) direction vector with (attacker > neighbor) vectors
- Additive bias instead of multiplicative (fixed offset regardless of defensibility), this should significantly mitigate the multiple border bias and directional paradoxes you mentioned. There's most likely edge cases though that could be problematic, maybe with disconnected attack fronts or nuked territory

concerns you mentioned:
- Topological mismatch: Still using euclidean geometry, but i think we avoid the problem since the priority queue only contains tiles that are 
1. on the active border
2. adjacent to conqurued territory
3. topologically reachable via the attack's expansion

The Euclidean calculation is only used for directional bias

late game instability: Since we're using normalized vectors (direction based), the spider like DFS pathology should be avoided.

The current centroid approach could still have unexpected behavior when attacking from multiple disconnected borders simultaneously, haven't tested this particular situation yet to be honest.

nottirb commented:
I think this is almost there, don't have time do a full analysis at the moment.

 Two things stand out:
 - Your current BFS implementation is essentially a more complex version of simple magnitude checks
 - You don't need to BFS the entire map

I'd suggest the following changes:
 - Restrict it to only the tiles owned by the player being attacked. That way the BFS distance is based on actually traversable tiles (you can't cross through water or your allies land during an attack).
 - This should also by proxy reduce the total number of tiles stored (since it will ignore tiles not owned by the player, and tiles disconnected from the clicked location).

Because the BFS portion is only *part* of your implementation, what you should see is that if more tiles become traversable than during the initial click (i.e: the person you're attacking is also attacking someone), it will be later in the attack and BFS will have a lower weight due to time decay.

There are also a bunch of optimization problems you can solve here once the core logic is done. If you could please send a gif update with how it looks once you adjust the BFS implementation that would be awesome. (It's hard to tell if certain paths are the wrong path since you're further than I got and everything I'm going off is just conceptual).

evanpelle commented:
> BFS added, had some issues with performance but is much better now. First try was doing the entire map with every attack, costing 310ms per attack :P
> 
> Could probably still be further optimized.
> 
> Tried on demand caching, amorlized, but once bots/players own large territories, it breaks down terribly.
> 
> Now is 10x downsized as suggested, could still be better probably. I didn't want to limit the depth lookup, which would have been easiest, but i wanted to avoid that as it may cause issues.
> 
> If it's good enough then might leave as, or optimize it further, maybe a hyprid approach.
> 
> edit: also centroid is now removed, and instead uses a per-tile normalized vector approach. All the above suggested by @nottirb - hopefully i understood it correctly. (thank you!)
> 
> edit 2: I left the performance measurements in the pull if anyone wants to test, can be seen after attacking in console > worker, even with chrome 6x CPU throttle, i cant spot any stutters or performance issues.
> 
> Still needs work for extreme use cases as suggested probably, and overall more comphrensive testing to see where it fails currently.

I recommend doing a chrome performance profile, and checking the worker thread (all these tile calculations happen in a worker thread)

MaxHT0x commented:
I changed some attack logic to match the original game system

**Before (Original main logic):** 
`const priority = defensibilityWeight + tickNow;`

**After (Reverted):**
```
const elapsedTicks = tickNow - this.attackStartTick;
const waveOffset = 40.0 * (1 - Math.exp(-elapsedTicks / 150.0));
let priority = defensibilityWeight + waveOffset;
```

**Impact:**
- **Old behavior:** Used absolute game time (tickNow), creating unbounded wave front growth
  - Short attacks, terrain = ~17% of priority calculation
  - Long attacks, terrain = ~0.4% of priority calculation (extreme example)
  - **Result: In large conquests, terrain becomes almost irrelevant**

- **New behavior (now reverted):** Wave front saturates at 40 points after ~450 ticks
  - All attacks: terrain consistently = ~33% of priority calculation
  - **Result: Terrain always matters ~equally, even in massive conquests**


I reverted the new behaviour to match the old system, since this would change balance and how attacks feel in general. 

New values are very very subtle, maybe too subtle but it's nice :D

MaxHT0x commented on tests/core/executions/DirectedAttack.test.ts:
```diff
@@ -0,0 +1,816 @@
+    expect(config.attackDirectionWeight()).toBe(2.5);
+    expect(config.attackTimeDecay()).toBe(20.0);
+    expect(config.attackMagnitudeWeight()).toBe(0.75);
+    expect(config.attackDistanceDecayConstant()).toBe(25.0);
```
can you apply it for me

evanpelle commented:
> PS. Im not sure the " Backward compatibility (attacks without clickTile)" means anything at all, since im not sure if previous replays are watchable once a major version is out, but i left the code in anyways. Let me know if it should be removed.

replays only work if the game record has the exact commit, so not point in worrying about backwards compatibility

evanpelle commented on src/core/execution/AttackExecution.ts:
```diff
@@ -291,6 +354,274 @@ export class AttackExecution implements Execution {
+   * @param downsampleFactor Sample tiles at grid coordinates (multiples of N)
+   * @returns Map of coarse tile refs to their BFS distance from click
+   */
+  private buildCoarseGrid(
```
Not sure if i'm understanding this right, so we just the BFS to create a distance map to avoid using manhattan dist? why is using manhattan distance a problem?

evanpelle commented on src/core/execution/AttackExecution.ts:
```diff
@@ -291,6 +354,274 @@ export class AttackExecution implements Execution {
+   * @param elapsedTicks Time since attack started (for decay factor)
+   * @returns Proximity bonus (negative value, or 0 if not available)
+   */
+  private calculateProximityBonus(
```
So this rewards tiles that are closer (according to bfs distance map) to the clicked tile?

MaxHT0x replied:
yes exactly, it also decays over time

evanpelle reviewed (commented):
It seems to me like we have two methods for calculating priority:

1. bfs distance
2. directionality

Do we need the bfs distance? Would this work with just directionality?

MaxHT0x commented:
> It seems to me like we have two methods for calculating priority:
> 
> 1. bfs distance
> 2. directionality
> 
> Do we need the bfs distance? Would this work with just directionality?


Technically yes if you set `attackMagnitudeWeight = 0.0`, which disables the proximity bonus entirely. BFS is only used for that. 

But we need both systems because they solve different problems. The directionality system works per-tile, each border tile calculates independently, which creates a rough directional bias with triangular convergence, but it's not precise about exact location. 

The proximity bonus (BFS) gives extra priority to tiles near the clicked location, and only tiles in the same connected region as the click receive this bonus (BFS only traverses defender's connected territory, water blocks the traversal). 

Example: Player A and Player B both have territory on the mainland plus an island separated by water. If Player A clicks to attack Player B's island, BFS will only track island tiles because water cuts it off. Island tiles receive a negative proximity bonus and are conquered first, while mainland tiles receive less proximity bonus and are focused more later as the decay happens. 

The attack will happen on both regions, but it strongly prioritizes the clicked island initially (island tiles get 20-70% priority boost depending on config).

Without BFS, both regions would have similar priority since both point "toward" the click geometrically.

MaxHT0x commented on src/core/configuration/DefaultConfig.ts:
```diff
@@ -803,6 +803,34 @@ export class DefaultConfig implements Config {
+
+  debugDirectedAttacks(): boolean {
+    return false;
+  }
```
No, this used to be the case but it was requested that the comments move somewhere else

PabloRuizCuevas commented:
imho, Sliders are a bit too complex, but directional attacks would be cool, maybe an like click-hold and drag in the direction you want to point.

So:
just click -> no directional (still very useful for multiple regions, even borders etc)
hold click -> an arrow appears, then we could chose the direction and strength by moving the head of the arrow and release it, kind of reverse bow gesture, or even bow gesture.

Just as idea, this interface could either control how directional it is the attack or how much troops to send.
